File: storage/guild_settings.py
"""
Per-guild persistent settings.

Keys stored per guild:
  volume     int   (0-200,  default 100)
  loop_mode  str   ("none"|"track"|"queue",  default "none")
  prefix     str   (optional per-guild prefix override)
"""
from __future__ import annotations

from typing import Any, Optional

from storage.store import JSONStore

_DEFAULTS: dict = {
    "volume": 100,
    "loop_mode": "none",
}


class GuildSettings:
    """Thin wrapper around JSONStore for per-guild settings."""

    def __init__(self, store: JSONStore) -> None:
        self._store = store

    # ──────────────────────────────────────────────────────────────────────────
    # Read
    # ──────────────────────────────────────────────────────────────────────────

    def get(self, guild_id: int, key: str, default: Any = None) -> Any:
        guild_data: dict = self._store.get(str(guild_id)) or {}
        if key in guild_data:
            return guild_data[key]
        if key in _DEFAULTS:
            return _DEFAULTS[key]
        return default

    def get_all(self, guild_id: int) -> dict:
        base = dict(_DEFAULTS)
        base.update(self._store.get(str(guild_id)) or {})
        return base

    # ──────────────────────────────────────────────────────────────────────────
    # Write
    # ──────────────────────────────────────────────────────────────────────────

    async def set(self, guild_id: int, key: str, value: Any) -> None:
        guild_data: dict = dict(self._store.get(str(guild_id)) or {})
        guild_data[key] = value
        await self._store.set(str(guild_id), guild_data)

    async def set_many(self, guild_id: int, **kwargs: Any) -> None:
        guild_data: dict = dict(self._store.get(str(guild_id)) or {})
        guild_data.update(kwargs)
        await self._store.set(str(guild_id), guild_data)

    async def reset(self, guild_id: int) -> None:
        await self._store.delete(str(guild_id))

    # ──────────────────────────────────────────────────────────────────────────
    # Convenience shortcuts
    # ──────────────────────────────────────────────────────────────────────────

    def volume(self, guild_id: int) -> int:
        return int(self.get(guild_id, "volume", 100))

    def loop_mode(self, guild_id: int) -> str:
        return str(self.get(guild_id, "loop_mode", "none"))

    def prefix(self, guild_id: int) -> Optional[str]:
        return self.get(guild_id, "prefix", None)

    async def save_volume(self, guild_id: int, volume: int) -> None:
        await self.set(guild_id, "volume", volume)

    async def save_loop_mode(self, guild_id: int, mode: str) -> None:
        await self.set(guild_id, "loop_mode", mode)

File: src/storage/guildSettings.js
/**
 * Per-guild persistent settings.
 *
 * Keys stored per guild:
 *   volume     number  (0-200, default 100)
 *   loop_mode  string  ("none"|"track"|"queue", default "none")
 *   prefix     string  (optional per-guild prefix override)
 */

const DEFAULTS = {
  volume: 100,
  loop_mode: 'none',
};

export class GuildSettings {
  /** @param {import('./store.js').JSONStore} store */
  constructor(store) {
    this._store = store;
  }

  // ── Read ───────────────────────────────────────────────────────────────────

  get(guildId, key, defaultValue = null) {
    const guildData = this._store.get(String(guildId)) || {};
    if (Object.prototype.hasOwnProperty.call(guildData, key)) return guildData[key];
    if (Object.prototype.hasOwnProperty.call(DEFAULTS, key)) return DEFAULTS[key];
    return defaultValue;
  }

  getAll(guildId) {
    return { ...DEFAULTS, ...(this._store.get(String(guildId)) || {}) };
  }

  // ── Write ──────────────────────────────────────────────────────────────────

  async set(guildId, key, value) {
    const guildData = { ...(this._store.get(String(guildId)) || {}) };
    guildData[key] = value;
    await this._store.set(String(guildId), guildData);
  }

  async setMany(guildId, updates) {
    const guildData = { ...(this._store.get(String(guildId)) || {}), ...updates };
    await this._store.set(String(guildId), guildData);
  }

  async reset(guildId) {
    await this._store.delete(String(guildId));
  }

  // ── Shortcuts ──────────────────────────────────────────────────────────────

  volume(guildId) {
    return Number(this.get(guildId, 'volume', 100));
  }

  loopMode(guildId) {
    return String(this.get(guildId, 'loop_mode', 'none'));
  }

  prefix(guildId) {
    return this.get(guildId, 'prefix', null);
  }

  async saveVolume(guildId, volume) {
    await this.set(guildId, 'volume', volume);
  }

  async saveLoopMode(guildId, mode) {
    await this.set(guildId, 'loop_mode', mode);
  }
}
